Reworking of image cache debug screen

amddg44 opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/414709148257752/1204422898101808/f
Tech Design URL:
CC:

**Description**:
Modification of the Image Cache screen to display all cached favicons to accurately track that deleting bookmarks, closing tabs etc clear cached favicons correctly

<!--
Tagging instructions
If this PR isn't ready to be merged for whatever reason it should be marked with the `DO NOT MERGE` label (particularly if it's a draft)
If it's pending Product Review/PFR, please add the `Pending Product Review` label.

If at any point it isn't actively being worked on/ready for review/otherwise moving forward (besides the above PR/PFR exception) strongly consider closing it (or not opening it in the first place). If you decide not to close it, make sure it's labelled to make it clear the PRs state and comment with more information.
-->

**Steps to test this PR**:

**Bookmarks:**
1. From a clean install, bookmark a few sites
2. Confirm favicons for these sites appear in the Image Cache screen in the Fireproof section tagged as `Bookmark`
3. Delete a bookmark and confirm the favicon is now gone
4. In `BookmarksViewController.swift` comment out [this line](https://github.com/duckduckgo/iOS/blob/fc2067847f2f062c9d70ef9b1ed9313ae86d9713/DuckDuckGo/BookmarksViewController.swift#L379). Run up the app and delete another bookmark. 
5. Confirm the favicon is now flagged as orphaned

**Fireproof Sites:**
1. In Settings enable `Ask to Fireproof on sign-in`
2. Fireproof a couple of sites
3. Confirm favicons for these sites appear in the Image Cache screen in the Fireproof section tagged as `Fireproof`
4. In `PreserveLoginsSettingsViewController.swift` comment out [this line](https://github.com/duckduckgo/iOS/blob/fc2067847f2f062c9d70ef9b1ed9313ae86d9713/DuckDuckGo/PreserveLoginsSettingsViewController.swift#L147). Run up the app and delete a fireproof site from the Fireproof Sites screen
5. Confirm the favicon is now flagged as orphaned  

**Autofill Logins:**
1. Enable `Save and Autofill Logins`
2. Visit some sites with logins and save a couple of autofill logins
3. Confirm favicons for these sites appear in the Image Cache screen in the Fireproof section tagged as `Login`
4. Delete a login and confirm the favicon is now gone
5. In AutofillLoginSettingsListViewController.swift comment out [these lines](https://github.com/duckduckgo/iOS/blob/fc2067847f2f062c9d70ef9b1ed9313ae86d9713/DuckDuckGo/AutofillLoginSettingsListViewController.swift#L257-L259). Run up the app and delete a login from the Logins screen. 
6. Confirm the favicon is now flagged as orphaned  

**General:**
1. Bookmark, fireproof and save an autofill login for the same site. Confirm favicons for these sites appear in the Image Cache screen in the Fireproof section tagged as `Bookmark`, `Login` and `Fireproof`
2. Confirm the favicons under the `Tabs` section match with your open tabs.
3. Close a couple of tabs and confirm that any closed tabs have their favicons flagged as orphaned 
4. Tap the fire button, and confirm the fireproof favicons are unchanged while the tab favicons have all been cleared

Editing to add screenshots! 

Fireproof favicions tagged with the reason why they are in the fireproof cache e.g. bookmark, login
<img width="202" alt="image" src="https://github.com/duckduckgo/iOS/assets/91189922/ee68ab03-fb37-4fbf-9299-1eac4cce6b94">
Example of flagged orphaned favicons:
<img width="202" alt="image" src="https://github.com/duckduckgo/iOS/assets/91189922/d8088603-bc8a-4b41-884f-f7a95c6bef64">


<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

amddg44 committed:
Re-working of ImageCacheDebugViewController to display all cached favicons and their sources

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/debug-image-cache-settings

samsymons reviewed (approved):
:+1: This new screen works great, nice work!

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/debug-image-cache-settings

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/debug-image-cache-settings

amddg44 committed:
Updated to use AutofillSecureVaultFactory

amddg44 committed:
Merge remote-tracking branch 'origin/develop' into anya/debug-image-cache-settings

amddg44 committed:
Keep Kingfisher tab cache in sync with open tabs (#1952)

Task/Issue URL: https://app.asana.com/0/414709148257752/1204375718000558/f
Tech Design URL:
CC:

Description:
One off cache clean-up of the tabs favicon cache + functionality added so that when a tab is closed, its favicon is deleted from tabs cache (as long as no other tabs have the same host)
